### ISSUE TITLE
Update `secrets edit` to dynamically find the preferred editor

### ISSUE THREAD

**Issue by aryanjassal** [MatrixAI/Polykey-CLI#277]
### Specification

This issue is a continuation of [https://github.com/MatrixAI/Polykey-CLI/issues/266](https://github.com/MatrixAI/Polykey-CLI/issues/266) ([ENG-398](https://linear.app/matrix-ai/issue/ENG-398/implement-secrets-edit-command)). In the previous issue, support for creating a file if it doesn't already exist in the vault was added. To finish off this issue, we need to identify the preferred editor of each platform and use that, instead of relying on `$EDITOR`.

Each platform has their own preferences for file editor. As such, identify and launch the corresponding file editor to edit the secret. Windows would have `notepad.exe`, and MacOS and Linux editors would require to be inferred from environment. To ensure everything is working smoothly, integration tests need to be done on all the given platforms.

> Generally speaking you rely on an environment variable: `$VISUAL` then if not set, rely on `$EDITOR`, and if not set, then you default onto something that would exist on most platforms.
>
> [https://unix.stackexchange.com/questions/4859/visual-vs-editor-what-s-the-difference](https://unix.stackexchange.com/questions/4859/visual-vs-editor-what-s-the-difference)
>
> The "default" depends on some magic.
>
> Here's our list depending on the OS.
>
> 1. Linux - VISUAL, EDITOR, `nano`, `ed` - I wouldn't default to vi or vim because it's actually a more advanced editor.
> 2. Windows - VISUAL, EDITOR, get the `.txt` file association first, then default to `notepad.exe` and wait for it to finish. Windows never had a default terminal text editor.
> 3. MacOS - VISUAL, EDITOR, `pico`, `nano`
>
> If the defaults cannot be found, you need to error out, indicating that no default editor could be found, and users can instead choose to set VISUAL, EDITOR or pass in the "editor" program with `--ed` or something.
>
> Furthermore it is important that the temporary being created can only be edited by the same user as the user who ran `polykey secrets edit`, otherwise it can be intercepted. The file should have a limited umask like `700`. Or equivalent for the Windows.

The behaviour of `secrets edit` command shouldn't be what is simplest. It should be what makes the most sense.

### Additional context

* Relates to #32 ([ENG-115](https://linear.app/matrix-ai/issue/ENG-115/incorporate-unix-commands-to-secrets-subcommand-pk-secrets))
* Preferred editor detection: [https://github.com/MatrixAI/Polykey-CLI/issues/32#issuecomment-2253699983](https://github.com/MatrixAI/Polykey-CLI/issues/32#issuecomment-2253699983)
* Continuation of [https://github.com/MatrixAI/Polykey-CLI/issues/266](https://github.com/MatrixAI/Polykey-CLI/issues/266) ([ENG-398](https://linear.app/matrix-ai/issue/ENG-398/implement-secrets-edit-command))

### Tasks

1. Rely on system editor to edit the secrets.
2. Test for desired behaviour on all platforms.

**Comment by aryanjassal** [comment]
This has already seen some work. The current default editor for Unix systems is `nano`. We try to get the preferred editor via the `$EDITOR` variable, and use the default when the variable is not available. In cases where `nano` is also not available, the command fails to spawn the child process and throws an error.

Perhaps a visual editor can be launched by a flag, like `polykey secrets edit vault:SECRET --visual`, which will attempt to launch the visual editor instead. As not all platforms have a default visual editor, or any visual editor at all, we can probably ignore providing a default for the visual.

Windows is being ignored for now, as the Windows build is currently defunct, and the `secrets edit` command hasn't even been implemented for Windows yet.

**Comment by CMCDragonkai** [comment]
What's going on with windows and is there an issue tracking that?

**Comment by CMCDragonkai** [comment]
Yea visual doesn't always exist on Linux, but there are defaults in Windows and MacOS because of the desktop. So we should auto find for MacOS and Windows.

**Comment by aryanjassal** [comment]
> What's going on with windows and is there an issue tracking that?

Yes, there is an issue tracking this. Windows support was never implemented for `secrets edit` command, and it cannot be tested either as Polykey CLI is currently broken on Windows anyways.

[https://github.com/MatrixAI/Polykey-CLI/issues/341](https://github.com/MatrixAI/Polykey-CLI/issues/341)<br>[https://github.com/MatrixAI/Polykey-CLI/issues/331](https://github.com/MatrixAI/Polykey-CLI/issues/331)

**Comment by aryanjassal** [comment]
I think the most elegant way to handle this issue is to have a priority list for each platform like so:

```ts
let editors: Array<string> = [];
switch (process.platform) {
  case 'linux': editors = [process.env.VISUAL, process.env.EDITOR, 'nano', 'ed']; break;
  case 'darwin': editors = [process.env.VISUAL, process.env.EDITOR, 'pico', 'nano']; break;
  default: break; // no windows support yet
}
editors.filter((editor) => Boolean(editor) && programExists(editor));
```

If `editors` is empty, no viable editor could be found. Otherwise, use the first available editor in the list. This would make editors easy to update.

I don't think we currently have a `programExists` function. The function needs to spawn an async child process which runs `command -v <program>` to check if it exists.
